aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/compute/strategy/CreateResourcesThenCreateNodes.java

```diff
@@ -0,0 +1,277 @@
+      PublicKey userKey = readPublicKey(options.getPublicKey());
+      final String fingerprint = computeFingerprint(userKey);
+      KeyPair keyPair;
+
+      synchronized (CreateResourcesThenCreateNodes.class) {
+         Optional<KeyPair> keyPairOptional = Iterables
+               .tryFind(api.sshKeyPairApi().list(regionId).concat(), new Predicate<KeyPair>() {
+                  @Override
+                  public boolean apply(KeyPair input) {
+                     return input.keyPairFingerPrint().equals(fingerprint.replace(":", ""));
+                  }
+               });
+         if (!keyPairOptional.isPresent()) {
+            logger.debug(">> key pair not found. Importing a new key pair %s ...", fingerprint);
+            keyPair = api.sshKeyPairApi().importKeyPair(
+                    regionId,
+                    options.getPublicKey(),
+                    namingConvention.create().uniqueNameForGroup(JCLOUDS_KEYPAIR_IMPORTED));
+            logger.debug(">> key pair imported! %s", keyPair);
+         } else {
+            logger.debug(">> key pair found for key %s", fingerprint);
+            keyPair = keyPairOptional.get();
+         }
+         return keyPair;
+      }
+   }
+
+   private KeyPairRequest generateKeyPair(String regionId, String prefix) {
+      logger.debug(">> creating default keypair for node...");
+      KeyPairRequest keyPairRequest = api.sshKeyPairApi().create(regionId, namingConvention.create().uniqueNameForGroup(prefix));
+      logger.debug(">> keypair created! %s", keyPairRequest);
+      return keyPairRequest;
+   }
+
+   private String getOrCreateSecurityGroupForOptions(String group, Location location,
+                                                 ECSServiceTemplateOptions options) {
+
+      checkArgument(options.getGroups().size() <= 1,
+              "Only one security group can be configured for each network interface");
+
+      String securityGroupId = null;
+      if (!options.getGroups().isEmpty()) {
+         Iterable<String> securityGroupNames = api.securityGroupApi().list(location.getId()).concat().transform(new Function<SecurityGroup, String>() {
+            @Override
+            public String apply(SecurityGroup input) {
+               return input.name();
+            }
+         });
+         for (String securityGroupName : options.getGroups()) {
+            checkState(Iterables.contains(securityGroupNames, securityGroupName), "Cannot find security group with name " + securityGroupName + ". \nSecurity groups available are: \n" + Iterables.toString(securityGroupNames)); // {
+         }
+      } else if (options.getInboundPorts().length > 0) {
+         String name = namingConvention.create().sharedNameForGroup(group);
+         SecurityGroupRequest securityGroupRequest = api.securityGroupApi().create(location.getId(),
+                 CreateSecurityGroupOptions.Builder.securityGroupName(name).vpcId(options.getVpcId()));
```

Comment:
The `vpcId` template option seems mandatory. LIke with the `vSwitch` one, options should not be mandatory. Can we provide a default value, or create one if no value has been provided?
At least, if this is not possible, we should validate the template options as a precondition to the `execute` method and fail soon with an understandable message.

Reply:
ok for validating them, not entirely sure about create a vSwitch if not available